src/blog/python-data-races.vmd

```diff
@@ -0,0 +1,146 @@
+ * Thread A stores register A into `counter`, it now contains 1.
+ * Thread B stores register B into `counter`, it now contains 1.
+
+As you can see, even though both threads incremented `counter`, it's not `2`, it's `1`. This is a data race in action.
+
+
+<slice />
+
+
+# The Intrigue
+
+
+This was actually pretty hard to discover. The first few experiments failed, because Python is pretty smart about when it runs each thread.
+
+
+Python will only interrupt a thread if it's taking too long. From [anekix's StackOverflow answer](https://stackoverflow.com/a/49573860):
+
+> In new versions, instead of using instruction count as a metric to switch threads, a configurable time interval is used. The default switch interval is 5 milliseconds.
+
+
+So, it seems a thread needs to last longer than 5 milliseconds to possibly trigger a data race. This explains why we had to do 100,000 iterations in each thread.
+
+
+This policy makes it difficult to identify when one has race conditions. However, it probably also reduces their frequency in production, which is a nice benefit.
+
+
+# Thoughts on Detecting Races
+
+
+As a language designer, I wonder if there was a missed opportunity in here.
+
+
+[Go's map iteration order is random](https://news.ycombinator.com/item?id=7655948), in part because it prevents us from accidentally relying on iteration order. We could take some inspiration from that.
+
+
+I wonder if, in development mode, Python could use a shorter interval so that we notice any race conditions hiding in our code, and in release mode, could use this longer (5ms) interval. [# We could do this today, by using `sys.setswitchinterval()`!]
+
+
+This also relates to one's philosophy on determinism. We know that:
+
+ * Determinism helps a lot for reproducing bugs. Nobody likes heisenbugs. [# A "heisenbug" is when we encounter a bug, but when investigating, have difficulty making it happen again, because it's based on some non-deterministic factor.]
+ * Determinism can cause some bugs. If we accidentally depend on a hash map's ordering in 100 places, and then change our hash map's algorithm, suddenly we have 100 bugs.
+
+
+These would seem to be incompatible, but a core goal of [Vale](https://vale.dev/) is to make races more obvious, and reproduce them easily.
+
+
+We could do this with *universal deterministic replayability*, where in development mode we record all non-deterministic inputs, such as command line arguments, stdin, sockets, files, etc. [# More specifically, it would just records any inputs from FFI into a file.] [# This feature is about halfway complete in Vale: the [FFI semantics](https://vale.dev/fearless) and FFI serialization are ready, but we don't record to a file yet.], plus the scheduling of all inter-thread messages and mutex lockings. This might seem difficult, but it's very possible when one uses a region-based borrow checker, and eliminates `unsafe` blocks and undefined behavior.
```

Comment:
Really it sounds like replayability is possible due to the language being deterministic. Stuff like the "region-based borrow checker" is how Vale maintains determinism, but there are other ways to do it. I would clarify you mean Vale is deterministic because of these such features.

Reply:
Changed to "This might seem difficult, but it's possible for a language to guarantee determinism, for example if it uses a"